### PR TITLE
gh-121249: Support _Complex types in the struct module

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-121249

----
📚 Documentation preview 📚: https://cpython-previews--121613.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/struct.rst [hunk 1]
```diff
@@ -267,12 +267,26 @@ platform-dependent.
 | ``P``  | :c:expr:`void \*`        | integer            |                | \(5)       |
 +--------+--------------------------+--------------------+----------------+------------+
 
+Additionally, if IEC 60559 compatible complex arithmetic (Annex G) is
```
Comment (ZeroIntensity) [inline Doc/library/struct.rst]: This should probably link to the specification.

File: Doc/library/struct.rst [hunk 2]
```diff
@@ -349,6 +363,11 @@ Notes:
    of bytes.  As a special case, ``'0s'`` means a single, empty string (while
    ``'0c'`` means 0 characters).
 
+(10)
+   For the ``'E'`` and ``'C'`` conversion codes, the packed representation uses
+   the IEEE 754 binary32 and binary64 format for components of the complex
```
Comment (ZeroIntensity) [inline Doc/library/struct.rst:368]: Again, link to the spec, we shouldn't expect people to have arbitrary specification numbers memorized.

File: Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Support :c:expr:`float complex` and :c:expr:`double complex`
+C types in the :mod:`struct` if compiler has C11 complex
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst]:
```suggestion
C types in the :mod:`struct` module if the compiler has C11 complex
```

File: Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Support :c:expr:`float complex` and :c:expr:`double complex`
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst]:
```suggestion
Support the :c:expr:`float complex` and :c:expr:`double complex`
```

File: Modules/_struct.c
Comment (ZeroIntensity) [inline Modules/_struct.c:1]: It's probably a better idea to define these functions even when `Py_HAVE_C_COMPLEX` isn't enabled, and then just raise a `struct.error` indicating that the system doesn't support it. I could see some people in the future being very confused as to why `struct` is complaining about an invalid format string across different systems.

**Review by ZeroIntensity — CHANGES_REQUESTED** [review]
Nice PR! This looks like something that would have had to get approved on Discourse, though.

File: Doc/library/struct.rst (hunk 1 above)
Comment (skirpichev) [reply Doc/library/struct.rst]: Which link do you suggest?  Wikipedia page for IEEE 754, C11 standard?

File: Doc/library/struct.rst (hunk 2 above)
Comment (skirpichev) [reply Doc/library/struct.rst:368]: I doubt it's arbitrary.  We have many references to the IEEE 754 standard in docs.

**Review by skirpichev — COMMENTED** [review]
> It's probably a better idea to define these functions even when Py_HAVE_C_COMPLEX isn't enabled, and then just raise a struct.error indicating that the system doesn't support it.

Patch follows to the ctypes module.  Not sure, maybe your approach is better.  Lets leave this to core devs.

**Comment by ZeroIntensity** [comment]
> Lets leave this to core devs.

This is why things get discussed on discourse first :wink:

File: Doc/library/struct.rst (hunk 1 above)
Comment (ZeroIntensity) [reply Doc/library/struct.rst]: Probably the C11 standard.

File: Doc/library/struct.rst [hunk 5]
```diff
@@ -267,12 +267,26 @@ platform-dependent.
 | ``P``  | :c:expr:`void \*`        | integer            |                | \(5)       |
 +--------+--------------------------+--------------------+----------------+------------+
 
+Additionally, if IEC 60559 compatible complex arithmetic (Annex G) is
+supported, the following format characters are available:
```
Comment (skirpichev) [inline Doc/library/struct.rst]:
```suggestion
Additionally, if IEC 60559 compatible complex arithmetic (Annex G of the
C11 standard) is supported, the following format characters are available:
```
?

File: Lib/test/test_struct.py [hunk 6]
```diff
@@ -34,6 +44,34 @@ def bigendian_to_native(value):
...
+            # both zero; check that signs match
+            elif math.copysign(1.0, x) == math.copysign(1.0, y):
+                return
+            else:
+                msg += ': zeros have different signs'
+        self.fail(msg.format(x, y))
+
+    def assertComplexesAreIdentical(self, x, y):
+        self.assertFloatsAreIdentical(x.real, y.real)
+        self.assertFloatsAreIdentical(x.imag, y.imag)
```
Comment (skirpichev) [inline Lib/test/test_struct.py]: These helpers refactored to a mixin class in https://github.com/python/cpython/pull/121071

File: Doc/library/struct.rst (hunk 5 above)
Comment (ZeroIntensity) [reply Doc/library/struct.rst]: That would be perfect.

**Review by ZeroIntensity — APPROVED** [review]
This implementation LGTM.

Although, for core dev review, I'm not too sure about this behavior: if the compiler does not support C11 complex types, then the new format characters are disabled and raise an error with no further information.

I'm worried that this will lead to some bug reports later on wondering why `E` and `C` don't work on their system -- it would be ideal to raise a different error mentioning that the system doesn't support C11 (albeit, it would be more work to implement).

File: Doc/library/struct.rst [hunk 7]
```diff
@@ -267,12 +267,26 @@ platform-dependent.
...
++--------+--------------------------+--------------------+----------------+------------+
+
 .. versionchanged:: 3.3
    Added support for the ``'n'`` and ``'N'`` formats.
 
 .. versionchanged:: 3.6
    Added support for the ``'e'`` format.
 
+.. versionchanged:: 3.14
+   Added support for ``'E'`` and ``'C'`` formats.
```
Comment (picnixz) [inline Doc/library/struct.rst]:
```suggestion
   Added support for the ``'E'`` and ``'C'`` formats.
```

To be consistent with the previous 'versionchanged' entries.

File: Doc/library/struct.rst [hunk 8]
```diff
@@ -349,6 +363,11 @@ Notes:
    of bytes.  As a special case, ``'0s'`` means a single, empty string (while
    ``'0c'`` means 0 characters).
 
+(10)
+   For the ``'E'`` and ``'C'`` conversion codes, the packed representation uses
```
Comment (picnixz) [inline Doc/library/struct.rst]:
```suggestion
   For the ``'E'`` and ``'C'`` format characters, the packed representation uses
```

I would use the same terminology as in the other notes.

**Review by picnixz — COMMENTED** [review]
Some minor comments but I also agree that having an error without info for an unknown format if complex are not supported is a bit weird.

**Comment by skirpichev** [comment]
Now on systems without _Complex type - a different error will be raised.

File: Modules/_struct.c
Comment (skirpichev) [reply Modules/_struct.c:1]: Done

File: Lib/test/test_struct.py [hunk 9]
```diff
@@ -774,6 +812,29 @@ def test_repr(self):
         s = struct.Struct('=i2H')
         self.assertEqual(repr(s), f'Struct({s.format!r})')
 
+    def test_c_complex_round_trip(self):
+        if not have_c_complex:
```
Comment (picnixz) [inline Lib/test/test_struct.py]: I would make two separate tests and use `skipIf` depending on the `have_c_complex` value. By the way, which build bots do not support float/double complex?

Comment (skirpichev) [reply Lib/test/test_struct.py]: > which build bots do not support float/double complex?

E.g. on Windows, WASI.

**Review by picnixz — APPROVED** [review]
On my side it's fine. I didn't check in details the functions bodies, but I assume that the tests should cover most of the cases. I think it remains to wait for Victor's and/or Serhiy's review.

**Comment by skirpichev** [comment]
As it's a follow up of https://github.com/python/cpython/pull/120894, CC @vstinner

File: Lib/test/test_struct.py [hunk 10]
```diff
@@ -783,6 +794,30 @@ def test_repr(self):
         s = struct.Struct('=i2H')
         self.assertEqual(repr(s), f'Struct({s.format!r})')
 
+    @unittest.skipUnless(have_c_complex, "requires C11 complex type")
```
Comment (vstinner) [inline Lib/test/test_struct.py]:
```suggestion
    @unittest.skipUnless(have_c_complex, "requires C11 complex type support")
```

File: Lib/test/test_struct.py [hunk 11]
```diff
@@ -783,6 +794,30 @@ def test_repr(self):
...
+    def test_c_complex_round_trip(self):
+        values = [complex(*_) for _ in combinations([1, -1, 0.0, -0.0, 2,
+                                                     -3, INF, -INF, NAN], 2)]
+        for z in values:
+            for f in ['E', 'C', '>E', '>C', '<E', '<C']:
+                with self.subTest(z=z, format=f):
+                    round_trip = struct.unpack(f, struct.pack(f, z))[0]
+                    self.assertComplexesAreIdentical(z, round_trip)
+
+    @unittest.skipIf(have_c_complex, "requires no C11 complex type")
```
Comment (vstinner) [inline Lib/test/test_struct.py]:
```suggestion
    @unittest.skipIf(have_c_complex, "requires no C11 complex type support")
```

File: Modules/_struct.c [hunk 12]
```diff
@@ -89,6 +96,10 @@ typedef struct { char c; _Bool x; } st_bool;
 #define LONG_ALIGN (sizeof(st_long) - sizeof(long))
 #define FLOAT_ALIGN (sizeof(st_float) - sizeof(float))
 #define DOUBLE_ALIGN (sizeof(st_double) - sizeof(double))
+#ifdef Py_HAVE_C_COMPLEX
+#define FLOAT_COMPLEX_ALIGN (sizeof(st_float_complex) - sizeof(float complex))
+#define DOUBLE_COMPLEX_ALIGN (sizeof(st_double_complex) - sizeof(double complex))
```
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
#  define FLOAT_COMPLEX_ALIGN (sizeof(st_float_complex) - sizeof(float complex))
#  define DOUBLE_COMPLEX_ALIGN (sizeof(st_double_complex) - sizeof(double complex))
```

File: Modules/_struct.c [hunk 13]
```diff
@@ -791,6 +822,58 @@ np_double(_structmodulestate *state, char *p, PyObject *v, const formatdef *f)
...
+{
+    Py_complex c = PyComplex_AsCComplex(v);
+    float complex x = CMPLXF((float)c.real, (float)c.imag);
+
+    if (c.real == -1 && PyErr_Occurred()) {
+        PyErr_SetString(state->StructError,
+                        "required argument is not a complex");
+        return -1;
+    }
+    memcpy(p, (char *)&x, sizeof(x));
```
Comment (vstinner) [inline Modules/_struct.c]: I'm not sure that the `(char *)` cast is useful. Same remark below.

File: Modules/_struct.c [hunk 14]
```diff
@@ -929,6 +1019,40 @@ bu_double(_structmodulestate *state, const char *p, const formatdef *f)
...
+    double x, y;
+
+    x = PyFloat_Unpack4(p, 0);
+    if (x == -1.0 && PyErr_Occurred()) {
+        return NULL;
+    }
+    y = PyFloat_Unpack4(p + 4, 0);
+    if (y == -1.0 && PyErr_Occurred()) {
+        return NULL;
+    }
```
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
    double x = PyFloat_Unpack4(p, 0);
    if (x == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
    double y = PyFloat_Unpack4(p + 4, 0);
    if (y == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
```

Same suggestion below.

File: Modules/_struct.c [hunk 15]
```diff
@@ -1197,6 +1360,40 @@ lu_double(_structmodulestate *state, const char *p, const formatdef *f)
...
+    double x, y;
+
+    x = PyFloat_Unpack4(p, 1);
+    if (x == -1.0 && PyErr_Occurred()) {
+        return NULL;
+    }
+    y = PyFloat_Unpack4(p + 4, 1);
+    if (y == -1.0 && PyErr_Occurred()) {
+        return NULL;
+    }
```
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
    double x = PyFloat_Unpack4(p, 1);
    if (x == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
    double y = PyFloat_Unpack4(p + 4, 1);
    if (y == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
```

Same suggestion below.

File: Modules/_struct.c [hunk 16]
```diff
@@ -1330,6 +1527,39 @@ lp_double(_structmodulestate *state, char *p, PyObject *v, const formatdef *f)
...
+    Py_complex x = PyComplex_AsCComplex(v);
+    if (x.real == -1 && PyErr_Occurred()) {
+        PyErr_SetString(state->StructError,
+                        "required argument is not a complex");
+        return -1;
+    }
+    if (PyFloat_Pack8(x.real, p, 1)) {
+        return -1;
+    }
+    return PyFloat_Pack8(x.imag, p + 4, 1);
```
Comment (vstinner) [inline Modules/_struct.c]: It shouldn't be p+8 here? If it's a real typo, I'm curious why it was missed by tests.

File: Modules/_struct.c (hunk 13 above)
Comment (skirpichev) [reply Modules/_struct.c]: I suspect it's useless as well.  Yet, it's used in all helpers, e.g. np_double right above.

Maybe clean up these casts across the file?

File: Modules/_struct.c (hunk 16 above)
Comment (skirpichev) [reply Modules/_struct.c]: I would guess that CI has only little-endian machines.  np_double_complex() was working instead.  Typo fixed.

File: Modules/_struct.c (hunk 13 above)
Comment (skirpichev) [reply Modules/_struct.c]: From the statistics, it seems that such remnants are mostly in this module:
```
$ git grep 'memcpy.*(char *\*)' | cut -f1 -d:|sort -n|uniq -c|sort -n
      1 Include/internal/pycore_call.h
      1 Modules/_ctypes/cfield.c
      1 Objects/call.c
      1 Objects/longobject.c
      1 Python/dtoa.c
      2 Modules/_io/bufferedio.c
      2 Objects/unicodeobject.c
      3 Modules/socketmodule.c
      5 Objects/memoryobject.c
     30 Modules/_struct.c
```

File: Modules/_struct.c (hunk 16 above)
Comment (vstinner) [reply Modules/_struct.c]: Oh right, `_structmodule_exec()` modifies `lilendian_table` at startup to use the native endian implementation instead. A big endian buildbot would catch the typo.

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Merged, thanks.

**Pull request merged**